### PR TITLE
[WIP; Crater] Use `parse_cond_expr` instead `parse_expr` in `if` guards

### PR DESCRIPTION
**Pull request by tesuji**
This PR improves diagnostic a bit for #15980.

### CODE REVIEW

**Comment by lzutao** [comment]
I added a test for unimplemented #51114 , also cc #74232 .

**Comment by estebank** [comment]
I think that the change is correct but I want a crater run in case this affects code out in the wild.

@bors try @craterbot check

**Review by estebank — APPROVED** [review]

**Comment by estebank** [comment]
@craterbot check

File: src/test/ui/rfc-2294-if-let-guard/non-implemented.rs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+// This function is to check whether `if-let` guard error points to
+// `let_chains` feature gate, which is wrong and confuse people.
+pub fn foo(a: &[u8], b: bool) -> bool {
+    match b {
+        true if let [1, 2, 3, ..] = a => true,
```
Comment (eddyb) [inline src/test/ui/rfc-2294-if-let-guard/non-implemented.rs:5]: But it *is* a subset of "`let` chains" (or rather, `let` expressions in more positions), at most you should change what feature-gate it points to.

Looking around at the let chains tests, you want all of these cases to emit feature-gate errors: https://github.com/rust-lang/rust/blob/6ee1b62c811a6eb68d6db6dfb91f66a49956749b/src/test/ui/rfc-2497-if-let-chains/feature-gate.rs#L111-L136

Note that `noop_expr!` and `#[cfg(FALSE)]` must emit feature-gate errors *despite* those AST nodes never reaching HIR lowering.

I think you should copy that entire test, remove `_while`, and then put every `if let` in a guard, e.g.:
```rust
match () { () if let ... = ... => {} }
```

**Comment by lzutao** [comment]
Closing in favor of #74566. 
Will triage crater there.

**Pull request closed without merging**